File: 2_last_digit_of_fibonacci_number/FibonacciLastDigit.kt
/*
 * Author: Jehadur Rahman
 * Organization: CyArm
 * GitHub: https://github.com/JehadurRE
 * Website: jehadurre.me
 * Date: 2025-07-10 09:43 (Thu)
 */

import java.util.*

fun getFibonacciLastDigit(n: Int): Int {
    if (n <= 1)
        return n

    var previous = 0
    var current = 1

    for (i in 0 until n - 1) {
        val tmpPrevious = previous
        previous = current
        current = (tmpPrevious + current) % 10
    }

    return current
}

fun main(args: Array<String>) {
    val scanner = Scanner(System.`in`)
    val n = scanner.nextInt()
    val c = getFibonacciLastDigit(n)
    println(c)
}

File: 2_last_digit_of_fibonacci_number/fibonacci_last_digit.hs
-- Author: Jehadur Rahman
-- Organization: CyArm
-- GitHub: https://github.com/JehadurRE
-- Website: jehadurre.me
-- Date: 2025-07-10 09:43 (Thu)

-- by Kirill Elagin

fibonacci_last_digit :: Int -> Int
fibonacci_last_digit = helper (0, 1)
  where
    helper (a, _) 0 = a `mod` 10
    helper (a, b) n = helper (b `mod` 10, (a + b) `mod` 10) (n - 1)

main :: IO ()
main = do
  [n] <- fmap words getLine
  print $ fibonacci_last_digit (read n)
